### ISSUE TITLE
Improve intentful-task workflow

### ISSUE THREAD

**Issue by shlomi-nx** [nutanix/papiea-js#337]
I am copying the checklist from #308, and adding more details. 

@Emulebest, Please update whats done, and whats next. You can break this into smaller chunks. If you do, please keep a reference here to the smaller tickets.

1. SDK - See the use-cases here https://github.com/nutanix/nunet-ui/issues/473
   1. [x] Add `on` function on a kind. 
        - [x] compile sfs and store in the kind in papiea-engine
   2. [x] Add `on_create` - Please see #339
   3. [x] Add `on_delete`
2. Engine - add intentful dispatch mechanism
    1. [x] Create a hook which invokes the handlers on creation events. Throwing an Error would prevent creation
    2. [x] Create a hook which invokes the handler on deletion events. Throwing an Error would prevent deletion. Any delete cascade logic could be handled here. 
    3. [ ] Handles spec changes:
         1. [x] When receiving a spec-change, create an Intentful Task 
         2. [ ] Add that entity to a "check diffs" list. every configurable duration check diff on every item in the list, if the minimum delay time is met
         3. [ ] a diff handler may respond back with a minimum delay time before next execution. If it does not, some default is used
         4. [ ] At some configurable time, the system will check diff on all other live entities (could do so on a random set of entities at every iteration). If a diff is found, the entity is added to "check diffs" list
        3. [x] Go over the "check diffs" list, if an entity has no diff, remove it from the list. If it has diffs, pick one at random (for now) and attempt to process. 
        4. [ ] When a handler is starting to work on a diff for an object, update a unique id of that handler in its task object
3. Intentful Tasks
    1. [x] Create a db collection for intentful tasks. It should store:
         1. a reference to the entity being changed
         2. some info on who and when made that change
         3. the list of diffs this task is tracking (all the diffs that got figured out within a SINGLE sfs function). It is described further in [this section of the design doc](https://nutanix.github.io/papiea-js/Papiea-design.html#h-Intentful-Tasks-117)
         4. an identifier into the handler function, so Papiea could check if the handler is still running or has terminated
    2. [ ] Monitor the progress of an Intentful task:
          - [ ] Essentially when all the fields that were supposed to change have either changed to the desired value (meaning a success) or changed to something else AND the spec version (and field's spec) have changed (meaning outdated).
          - [ ] This should be tracked by the update spec/ update status functions in papiea

**Comment by shlomi-nx** [comment]
Currently intentful tasks are a bit hard to follow and convoluted. 

e.g. When attempting to get this into a working state, I ended up clearing "done" tasks in `checkActiveTasksHealth` function, which may not be correct:

https://github.com/nutanix/papiea-js/blob/89be5443de0d6378fa3a1e44b2bc5428833b772b/papiea-engine/src/tasks/differ_resolver.ts#L196-L202

Lets try to make this a bit more clear:

Main loop:
1. For each entity `e` in "check diff" list
   1. Is entity minimum delay time met? 
   2. No - continue to next entity
   3. Yes - Is there a task `t` working on `e`?
      1. Yes - If the task is active, continue to next entity
      2. No - Check for diffs
         1. No diffs found - remove entity from "check diff" list
         2. Diff found, pick most concrete SFS (if few are at the same level, pick at random) and execute the handler
         3. When handler returns with a "minimum delay time" make a note when this entity needs to rediff again. If it does not return with a "minimum delay time" assume a default value was returned
         
Spec/Status updates on entity `e`:
1. Add entity to "check diff" list
2. Check if there is a task `t` with state `pending` or `active` on entity `e`:
    1. Yes:
         1. Check all the fields being watched
             1. Implement field logic as described [here](https://nutanix.github.io/papiea-js/Papiea-design.html#h-Intentful-Tasks-117)
             2. set the task to the correct state depending on the different tracked fields

**Comment by Emulebest** [comment]
I updated the state of things ATM. I only ticked those that are proven working by tests and don't have any unnecessary complexity about them, all the others need some adjustments.

@shlomi-nx have a question, what is regarded as `minimum delay time` and what is the lifecycle of that? Who should set it? When?

**Comment by shlomi-nx** [comment]
Think of what we currently have:
https://github.com/nutanix/papiea-js/blob/9ded5b7100563d28878f647b78f07e86595c6652/papiea-engine/src/tasks/differ_resolver.ts#L43-L50

This `delay` is global. What it means is that every `delay` time, all the operations would be attempted to run, and then nothing runs for another `delay` duration. Practically, if a handler fails to succeed, it will get reinvoked at the end of the global `delay` duration.

This is a problem: 
1. Some operation may want to be invoked sooner/later than the global delay
2. The system will have time when nothing happens, and spikes of activity every `delay` duration

The concept of the `minimum delay time` is that instead of using a global delay, each diff can use a local delay time, with some default.

This means that when a new diff is found, its delay is zero, so the next time the system woke up, it tries to handle it, as well as all the other diffs which their local delay has expired. When a handler fails, it can say what is the minimum delay it wants to wait before getting reinvoked.

Using this method, the global delay can be rather small, say 300 ms, but not all diffs will get executed at each run, only new ones and ones that has expired. 

Example:
1. t=0ms: Identified diff A, add to watch list with delay 0
2. t=500ms: resolver loop woke up from global delay (500 ms), found diff A, executed it
3. t=600ms: A execution failed, with a delay of 1 second (will expire at 1600ms)
4. t=700ms: Identified diff B, add to watch list with delay 0
5. t=1000ms: resolver loop woke up from global delay (500 ms), diff A not expired yet, found diff B, executed it which takes 1 seconds to execute 
6. t=1500ms: resolver loop woke up from global delay (500 ms), diff A not expired yet, diff B still working, nothing to do
7. t=2000ms: resolver loop woke up from global delay (500 ms), diff A has expired, re-executing it
8. t=2000ms: B failed with minimum delay 200ms (will expire at 2200ms)
9. t=2100ms: A resolved successfully
10. t=2500ms: resolver loop woke up from global delay (500 ms), diff A has been resolved, removing from watch, B has expired, re-executing it

**Comment by adolgarev** [comment]
@Emulebest this is like https://erlang.org/doc/design_principles/sup_princ.html#tuning-the-intensity-and-period or https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/429

**Comment by Emulebest** [comment]
Linking this to #356

**Comment by shlomi-nx** [comment]
@Emulebest, Is this done? Please link PR and close. If there are some smaller things to finalize, lets open smaller tickets to track those.

**Comment by Emulebest** [comment]
resolved by #356
